### PR TITLE
[CBRD-23494] New page buffer and DWB stats

### PR DESCRIPTION
**Pull request by ribram**
http://jira.cubrid.org/browse/CBRD-23494

There are some changes required to DWB statistics:

  - remove "Time" prefix/suffix from names; counter/timers are automatically prefixed when printed.
  - Time_DWB_checksum_time
  - I need help describing Time_DWB_flush_block_helper_time, Time_DWB_wait_flush_block_time, Time_DWB_wait_flush_block_helper_time. Might be better to also have a more suggestive name for "helper"
  - Time_DWB_flush_block_cond_wait_time may also be renamed better.

@alexstanro what do you think?

### CODE REVIEW

**Comment by alexstanro** [comment]
1. I agree to remove time prefix.

2. Time_DWB_checksum_time. The purpose was to measure checksum computation time. Initially, it was an expensive job, so statistics was mandatory. Since now checksum computation is faster, we can safely remove this statistics.

3. In most of the cases there are 2 threads involved, as follow:
- flush block daemon - used to write & sync DWB volume,  to write & sync in data volume, manage internal structures (requiring synchronization) to optimize flush. Since this thread does expensive jobs, data volumes may be sync by other thread (flush helper block daemon).
- flush helper block daemon used to speed up flush. Allows to sync a volume, while flush block daemon does other expensive job (sync another volume, write, manage internal structures that require synchronization)

- Time_DWB_flush_block_helper_time - measures the time spent by flush helper block daemon
- Time_DWB_wait_flush_block_time - measures the time while a thread can't add a new page to DWB in memory, because the flush of that memory was not finalized yet.
- Time_DWB_wait_flush_block_time - In case that flush helper block daemon didn't finished to sync previous block (should happens rare), flush block daemon must wait before flushing current block (flushing the blocks in order must be guarantee). This statistics measure this wait.

I don't know a better name for now. Renaming stats will also will requires also renaming some functions, threads...  It is more important to identify all possible bottlenecks and to have statistics.

File: en/admin/admin_utils.rst [hunk 1]
```diff
@@ -2326,6 +2339,30 @@ The following shows [options] available with the **cubrid statdump** utility.
...
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_flush_block_helper_time       | Counter/timer  | ??                                                                    |
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_flush_block_cond_wait_time    | Counter/timer  | The number and duration of DWB thread sleeps                          |
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_flush_block_sort_time         | Counter/timer  | The number and duration of sorting pages for flush                    |
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_flush_remove_hash_entries     | Counter/timer  | The number and duration of removing page from DWB                     |
```
Comment (alexstanro) [inline en/admin/admin_utils.rst]: Used to remove all pages ids of latest written block from hash.

File: en/admin/admin_utils.rst [hunk 2]
```diff
@@ -2326,6 +2339,30 @@ The following shows [options] available with the **cubrid statdump** utility.
...
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_checksum_time                 | Counter/timer  | !!Unused                                                              |
```
Comment (alexstanro) [inline en/admin/admin_utils.rst]: It is obsolete, it can be removed.

File: en/admin/admin_utils.rst [hunk 3]
```diff
@@ -2326,6 +2339,30 @@ The following shows [options] available with the **cubrid statdump** utility.
...
     +------------------------------------------+----------------+-----------------------------------------------------------------------+
 
+    **Double write buffer**
+
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | **Stat name**                            | **Stat type**  |  **Description**                                                      |
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_flush_block_time              | Counter/timer  | The number and duration of DWB block flushes                          |
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_flush_block_helper_time       | Counter/timer  | ??                                                                    |
```
Comment (alexstanro) [inline en/admin/admin_utils.rst]: Maybe volume prefix may be added. DWB_flush_block_volume_time for flush block daemon, Time_DWB_flush_block_volume_helper_time for flush block helper daemon.

File: en/admin/admin_utils.rst [hunk 4]
```diff
@@ -2326,6 +2339,30 @@ The following shows [options] available with the **cubrid statdump** utility.
...
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_flush_block_sort_time         | Counter/timer  | The number and duration of sorting pages for flush                    |
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_flush_remove_hash_entries     | Counter/timer  | The number and duration of removing page from DWB                     |
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_checksum_time                 | Counter/timer  | !!Unused                                                              |
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_wait_flush_block_time         | Counter/timer  | ??                                                                    |
```
Comment (alexstanro) [inline en/admin/admin_utils.rst]: Waits to add a page in a full block, not flushed yet.

File: en/admin/admin_utils.rst [hunk 5]
```diff
@@ -2326,6 +2339,30 @@ The following shows [options] available with the **cubrid statdump** utility.
...
+    +------------------------------------------+----------------+-----------------------------------------------------------------------+
+    | ..Time_DWB_wait_flush_block_helper_time  | Counter/timer  | ??                                                                    |
```
Comment (alexstanro) [inline en/admin/admin_utils.rst]: flush block daemon must waits for previous block to be completely flushed, before flushing the current block.

**Comment by ribram** [comment]
Summary of DWB statistics changes:

  1. Remove "time" prefix/suffix from names
  2. Remove DWB_checksum stat
  3. Rename DWB_flush_remove_hash_entries as DWB_remove_flushed_pages (the manual description should also be updated)
  4. Give "helper" a more suggestive name, understandable by user. file_sync_helper, separate_file_sync? any other suggestions? For description, we may say "number and duration of file synchronization by DWB helper" or "number and duration of separate file synchronization".
  5. DWB_wait_flush_block => can we add prev(ious)? e.g. DWB_wait_flush_prev(ious)_block

@alexstanro please check and confirm or add other suggestions.

@eseokoh do you want to add something else?

**Comment by alexstanro** [comment]
3. remove_flushed is unapropriate
4. I agree with file_sync_helper and "number and duration of file synchronization by DWB helper"
5. No. Previous block may means different block. In fact, we are waiting for previous version of same block...
6. Yes.

**Pull request merged**